### PR TITLE
Add functionality to log in using uni token and implement priorities endpoint

### PR DESCRIPTION
**Pull request by fabianfreyer**
This PR will be continuously updated as the token login gets developed.

### CODE REVIEW

File: app/registration/__init__.py [hunk 1]
```diff
@@ -1,8 +1,16 @@
-from flask import Blueprint
+from flask import Blueprint, g
```
Comment (JanLuca) [inline app/registration/__init__.py]: What is "g"? I don't find a documentation...

Comment (fabianfreyer) [inline app/registration/__init__.py]: http://flask.pocoo.org/docs/0.12/api/#flask.g

**Comment by JanLuca** [comment]
@fabianfreyer : Please check the commit I pushed to the PR

File: app/registration/models.py [hunk 2]
```diff
@@ -10,6 +10,9 @@ class Registration(db.Model):
     confirmed = db.Column(db.Boolean())
     uni_id = db.Column(db.Integer(), db.ForeignKey('uni.id'))
     uni = db.relationship('Uni', backref=db.backref('Registrations', lazy='dynamic', cascade="all, delete-orphan"))
+    uni_order = db.Column(db.Integer())
```
Comment (fabianfreyer) [inline app/registration/models.py]: we have the priority field for this.

File: app/registration/api.py [hunk 3]
```diff
@@ -44,7 +44,39 @@ def api_register():
...
+                registration.uni_order = req['order'].index(reg_id)
+            except ValueError:
+                registration.uni_order = new_order
+                new_order = new_order + 1
+            db.session.add(registration)
+
+        db.session.commit()
+        return "OK"
+
+    registrations = Registration.query.filter_by(uni_id = g.uni.id).order_by(Registration.uni_order)
```
Comment (fabianfreyer) [inline app/registration/api.py]: should be `.priority`

File: app/registration/api.py [hunk 4]
```diff
@@ -44,7 +44,39 @@ def api_register():
         data = registration.blob,
     )
 
-@registration_blueprint.route('/api/order/token')
+@registration_blueprint.route('/api/order', methods=['GET', 'POST'])
+@oauth.require_oauth('uni_order')
 @token_auth.login_required
-def validate_token():
-    return jsonify(g.uni.name)
+def api_registration_order():
```
Comment (fabianfreyer) [inline app/registration/api.py]: I'd make this `api_registration_priorities()`

File: app/registration/api.py [hunk 5]
```diff
@@ -44,7 +44,39 @@ def api_register():
         data = registration.blob,
     )
 
-@registration_blueprint.route('/api/order/token')
+@registration_blueprint.route('/api/order', methods=['GET', 'POST'])
+@oauth.require_oauth('uni_order')
 @token_auth.login_required
-def validate_token():
```
Comment (fabianfreyer) [inline app/registration/api.py]: Why is this being deleted?

File: app/registration/api.py [hunk 6]
```diff
@@ -44,7 +44,39 @@ def api_register():
...
-    return jsonify(g.uni.name)
+def api_registration_order():
+    if request.method == 'POST' \
+        and request.headers.get('Content-Type') == 'application/json':
+        req = request.get_json()
+        if not req:
+            return '', 403
```
Comment (fabianfreyer) [inline app/registration/api.py]: `return req or abort(403)`

File: app/registration/api.py [hunk 7]
```diff
@@ -44,7 +44,39 @@ def api_register():
         data = registration.blob,
     )
 
-@registration_blueprint.route('/api/order/token')
+@registration_blueprint.route('/api/order', methods=['GET', 'POST'])
+@oauth.require_oauth('uni_order')
```
Comment (fabianfreyer) [inline app/registration/api.py]: ich finde `registration_priorities` passender, da es ja nicht die Reihenfolge der Unis ist.

File: app/registration/api.py [hunk 8]
```diff
@@ -44,7 +44,39 @@ def api_register():
...
+        for reg_id in req['confirmed']:
+            registration = Registration.query.filter_by(id = reg_id).first()
+            registration.confirmed = True
+            try:
+                registration.uni_order = req['order'].index(reg_id)
+            except ValueError:
+                registration.uni_order = new_order
+                new_order = new_order + 1
+            db.session.add(registration)
+
```
Comment (fabianfreyer) [inline app/registration/api.py]: Missing logic to revoke confirmation.

File: app/registration/api.py [hunk 9]
```diff
@@ -44,7 +44,39 @@ def api_register():
...
+            return '', 403
+
+        new_order = len(req['order'])
+
+        for reg_id in req['confirmed']:
+            registration = Registration.query.filter_by(id = reg_id).first()
+            registration.confirmed = True
+            try:
+                registration.uni_order = req['order'].index(reg_id)
+            except ValueError:
```
Comment (fabianfreyer) [inline app/registration/api.py]: This `try`...`catch` block is confusing. May I suggest `enumerate(req['confirmed'])`?

File: app/registration/api.py [hunk 10]
```diff
@@ -44,7 +44,39 @@ def api_register():
...
+                registration.uni_order = new_order
+                new_order = new_order + 1
+            db.session.add(registration)
+
+        db.session.commit()
+        return "OK"
+
+    registrations = Registration.query.filter_by(uni_id = g.uni.id).order_by(Registration.uni_order)
+    confirmed = []
```
Comment (fabianfreyer) [inline app/registration/api.py]: This should be a list comprehension:

```python
confirmed = [
   {'reg_id': reg.id, 'name': reg.user.full_name, 'mail': reg.user.mail, 'order': reg.uni_order}
   for reg in registrations if reg.confirmed
  ]
not_confirmed = [
   {'reg_id': reg.id, 'name': reg.user.full_name, 'mail': reg.user.mail, 'order': reg.uni_order}
   for reg in registrations if not reg.confirmed
  ]
```

File: app/registration/models.py (hunk 2 above)
Comment (JanLuca) [inline app/registration/models.py]: Ok, I will change this

File: app/registration/api.py (hunk 7 above)
Comment (JanLuca) [inline app/registration/api.py]: Ok, will be fixed

File: app/registration/api.py (hunk 5 above)
Comment (JanLuca) [inline app/registration/api.py]: This is just a rename of the function

File: app/registration/api.py (hunk 4 above)
Comment (JanLuca) [inline app/registration/api.py]: Ok, will be changed

File: app/registration/api.py (hunk 6 above)
Comment (JanLuca) [inline app/registration/api.py]: Ok, this is a better style...

File: app/registration/api.py (hunk 9 above)
Comment (JanLuca) [inline app/registration/api.py]: But the enumerate(req['confirmed']) is not very obvious, too. Maybe another idea?

File: app/registration/api.py (hunk 8 above)
Comment (JanLuca) [inline app/registration/api.py]: Will be added later, the PR is WIP ;)

File: app/registration/api.py (hunk 10 above)
Comment (JanLuca) [inline app/registration/api.py]: Yeah, better style, you are right

**Comment by JanLuca** [comment]
@fabianfreyer : New version, please review

File: app/registration/api.py [hunk 11]
```diff
@@ -44,7 +44,44 @@ def api_register():
...
+
+        new_priority = len(req['priority'])
+
+        registrations = Registration.query.filter_by(uni_id = g.uni.id)
+        for registration in registrations:
+            if registration.id in req['confirmed']:
+                registration.confirmed = True
+                try:
+                    registration.priority = req['order'].index(reg_id)
+                except ValueError:
```
Comment (fabianfreyer) [inline app/registration/api.py]: Why exactly is there a `ValueError` thrown here? What case does this `try`...`except` block here?

Comment (JanLuca) [inline app/registration/api.py]: Adding a comment is a good point, I will add it

**Comment by JanLuca** [comment]
@fabianfreyer : Please review the PR again

File: app/registration/api.py [hunk 12]
```diff
@@ -44,7 +44,50 @@ def api_register():
...
+            if registration.id in req['confirmed']:
+                registration.confirmed = True
+                # The try-catch block checks if the id of the 
+                # registration is in the req['order'] list.
+                # If found the index of the element in the list will be used
+                # as priority of the registration.
+                # If not found the registration will get the next free
+                # priority as default value.
+                try:
+                    registration.priority = req['order'].index(reg_id)
```
Comment (fabianfreyer) [inline app/registration/api.py]: `reg_id` is not defined

File: app/registration/api.py [hunk 13]
```diff
@@ -44,7 +44,50 @@ def api_register():
...
+        and request.headers.get('Content-Type') == 'application/json':
+        req = request.get_json()
+        if not req:
+            abort(403)
+            return ''
+
+        new_priority = len(req['priority'])
+
+        registrations = Registration.query.filter_by(uni_id = g.uni.id)
+        for registration in registrations:
```
Comment (fabianfreyer) [inline app/registration/api.py]: TBH, I still don't see how this cannot just be:
```python
# Get a list of all registrations by this uni
registrations = {reg.id: reg for reg in Registration.query.filter_by(uni_id = g.uni.id)}

# Set the priorities according to the order they are given in in the confirmed list
for priority, reg_id in enumerate(req['confirmed']):
    reg = registrations.pop(reg_id, None)
    # FIXME: error handling for the case reg == none
    reg.confirmed = True
    reg.priority = priority
    db.session.add(reg)

# Set all remaining priorities as unconfirmed
for reg in registrations:
    reg.confirmed = False
    reg.priority = None
    db.session.add(reg)
```

Additionally, we might want to add a check for negative priorities (guaranteed places).

**Comment by fabianfreyer** [comment]
Also, please revert 137eb6031504f556126d9b4f7382e8ff631587de. We don't require anyone to be logged in using OAuth (since that's how you get to the registration form), so token auth **should** replace OAuth here.

File: app/registration/api.py (hunk 13 above)
Comment (JanLuca) [inline app/registration/api.py]: I have no problem to use your solution, IMHO this is more a question of style ;)

Anyway, at the moment we do not have any information about guaranteed places, I do not like the idea to symbolize them with negative priorities...

**Comment by JanLuca** [comment]
I prefer the second authorization with OAuth because we can use the communication interface already used in the Anmeldesystem with the oauth-GET/POST methods. This ensure that the API can only be used by applications with the right scope and valid callback URLs. So I do not want to 137eb60

**Comment by fabianfreyer** [comment]
That means that someone would have to login with their username/password to access the API, no?

**Comment by JanLuca** [comment]
Yeah, you have to login to access the priority API, that is right

File: app/registration/models.py [hunk 14]
```diff
@@ -6,11 +6,13 @@ class Registration(db.Model):
...
     username = db.Column(db.Text(), unique = True)
     blob = db.Column(db.Text())
-    priority = db.Column(db.Integer(), unique = True)
+    priority = db.Column(db.Integer())
     confirmed = db.Column(db.Boolean())
     uni_id = db.Column(db.Integer(), db.ForeignKey('uni.id'))
     uni = db.relationship('Uni', backref=db.backref('Registrations', lazy='dynamic', cascade="all, delete-orphan"))
 
+    priority_constraint = db.UniqueConstraint('uni_id', 'priority', name='uni_priority_unique')
+
```
Comment (fabianfreyer) [inline app/registration/models.py]: We will probably need a migration here. I haven't yet set up migrations... I guess I should?

Comment (fabianfreyer) [inline app/registration/models.py]: After #24 is merged, I'll rebase this on master and generate a migration for the unique constraint here.

**Pull request merged**